policy_agent/policy_agent.py

```diff
@@ -0,0 +1,643 @@
+
+
+class PolicyAgent():
+
+    """
+    The Policy Agent is responsible for maintaining Watch Threads/Queues and internal resource lists
+    """
+
+    def __init__(self):
+        self.q = Queue.Queue()
+        self.namespaces = dict()
+        self.services = dict()
+        self.pods = dict()
+        self.endpoints = dict()
+
+        self.changed_namespaces = dict()
+        self.changed_services = dict()
+        self.changed_pods = dict()
+        self.changed_endpoints = dict()
+
+    def run(self):
+        """
+        PolicyAgent.run() is called at program init to spawn watch threads and parse their responses
+        """
+        self.init_lists()
+        self.init_threads()
+
+        self.PodWatcher.start()
+        self.SvcWatcher.start()
+        self.NsWatcher.start()
+        self.EptsWatcher.start()
+
+        while True:
+            self.read_responses()
+
+    def init_lists(self):
+        _log.info("Initializing Resource Lists")
+        initNamespaces = _get_api_list("namespaces")
+        initServices = _get_api_list("services")
+        initPods = _get_api_list("pods")
+        initEndpoints = _get_api_list("endpoints")
+
+        for ns in initNamespaces["items"]:
+            self.process_resource(action="ADDED", kind="Namespace", target=ns)
+
+        for svc in initServices["items"]:
+            self.process_resource(action="ADDED", kind="Service", target=svc)
+
+        for pod in initPods["items"]:
+            self.process_resource(action="ADDED", kind="Pod", target=pod)
+
+        for ep in initEndpoints["items"]:
+            self.process_resource(action="ADDED", kind="Endpoints", target=ep)
+
+        self.nsRV = initEndpoints["metadata"]["resourceVersion"]
+        self.svcRV = initServices["metadata"]["resourceVersion"]
+        self.poRV = initPods["metadata"]["resourceVersion"]
+        self.epRV = initNamespaces["metadata"]["resourceVersion"]
+
+        _log.info("Service Resource Version = %s" % self.svcRV)
+        _log.info("Namespace Resource Version = %s" % self.nsRV)
+        _log.info("Pod Resource Version = %s" % self.poRV)
+        _log.info("Endpoints Resource Version = %s" % self.epRV)
+
+    def init_threads(self):
+        self.NsWatcher = Thread(target=_keep_watch,
+                                args=(self.q, "namespaces", self.nsRV))
+        self.SvcWatcher = Thread(target=_keep_watch,
+                                 args=(self.q, "services", self.svcRV))
+        self.PodWatcher = Thread(target=_keep_watch,
+                                 args=(self.q, "pods", self.poRV))
+        self.EptsWatcher = Thread(target=_keep_watch,
+                                  args=(self.q, "endpoints", self.epRV))
+
+        self.NsWatcher.daemon = True
+        self.SvcWatcher.daemon = True
+        self.PodWatcher.daemon = True
+        self.EptsWatcher.daemon = True
+
+    def read_responses(self):
+        """
+        Read Responses pulls a response off the Queue and processes it
```

Comment:
"pulls a response" -- what is a "response" in this case?

I think this method is actually "read_updates" - we aren't querying anything, so we're not receiving responses to a query.  We've just registered for updates.